### PR TITLE
Add simple asynchronous processing example

### PR DESCRIPTION
**Pull request by cmichi**
This is a simple example which reads from stdin line by line, an asynchronous operation is applied to each line, and the output is written to stdout asynchronously.

I think this example could help people get started with the lib. From my point of view this is a common use case ‒ at least it is was a scenario that I needed the lib for and I had similar use cases in the past.                                                                                                             

If you decide against merging it, I would still be interested in feedback if this is the idiomatic way to use the lib.

This is how the example behaves:
```
$ yes foobar | head -n5 | cargo run --example line-by-line
    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
     Running `target/debug/examples/line-by-line`                                                                                                                        
FOOBAR                                                                                                                                                                   
FOOBAR
FOOBAR
FOOBAR
FOOBAR
```

### CODE REVIEW

File: examples/line-by-line.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
+extern crate tokio_stdin_stdout;
+extern crate tokio;
+extern crate tokio_core;
```
Comment (vi) [inline examples/line-by-line.rs]: What is the reason of including both `tokio` and `tokio-core`?

File: examples/line-by-line.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+fn main() {
+  let mut stdin = tokio_stdin_stdout::stdin(0);
+  let stdout = tokio_stdin_stdout::stdout(0).make_sendable();
+
+  let mut current_line = String::from("");
+
+  loop {
+    let buf = vec![0; 1];
```
Comment (vi) [inline examples/line-by-line.rs]: One byte at a time looks inefficient. Maybe there should be `tokio-codec` for processing things linewise or something?

**Comment by vi** [comment]
The library expects users to already know what to do with `AsyncRead` and `AsyncWrite`. But more examples won't hurt anyway.

Is the reason of this example to show how to process text input line by line or to process binary input chunk by chunk?

**Comment by cmichi** [comment]
I've amended the commit based on your remarks. The example is now smaller and better comprehensible, thanks!

> Is the reason of this example to show how to process text input line by line or to process binary input chunk by chunk?

The idea was to process text input line by line and convert it to uppercase, so I think the `LinesCodec` works well now. What would you have done differently if the input were binary? Just the `BytesCodec`?

**Comment by vi** [comment]
> What would you have done differently if the input were binary? Just the BytesCodec?

Maybe `BytesCodec`, maybe without tokio-codec at all. `LinesCodec` on binary data means buffering chunks in memory arbitrarily, hoping for occasional `\n`.

**Comment by vi** [comment]
I'm going to accept this, but with some significant changes (a comment will follow).

**Comment by vi** [comment]
Changes:

1. Fix error - only one line were processed (950da81832cb1dd42981935852bdc7f5844e4d5a)
2. Make use of `FramedWrite` as well
3. <s>Added a final message</s>
4. Added more comments
5. Changed it to single-threaded Tokio mode (there is already unual `tokio::run` in neighbouring example), also with comments.
6. Listed examples in README.

**Comment by vi** [comment]
Do you like the state of the example as of now?

**Comment by vi** [comment]
Added [async fn version of this example](https://github.com/vi/tokio-stdin-stdout/blob/master/examples2/linebyline_asyncfn1.rs).

**Comment by cmichi** [comment]
Ah, very nice! Yeah the examples are great. Thanks for the improvements, I learned new things from them :+1:!

**Pull request merged**